habitat-lab/habitat/tasks/rearrange/multi_task/composite_sensors.py

```diff
@@ -108,52 +125,55 @@ def update_metric(self, *args, episode, task, observations, **kwargs):
+        # (for example the agent hasn't called terminate) we give zero reward
 
-        to_goal = task.measurements.measures[
-            ObjectToGoalDistance.cls_uuid
-        ].get_metric()
-        to_obj = task.measurements.measures[
-            EndEffectorToObjectDistance.cls_uuid
-        ].get_metric()
+        if self._cur_rearrange_stage == self.num_targets:
+            self._metric = 0
+            return
+
+        obj_to_goal_dist = self.get_distance(task, ObjectToGoalDistance)
+        ee_to_obj_dist = self.get_distance(task, EndEffectorToObjectDistance)
+
+        is_holding_obj = self._sim.grasp_mgr.snap_idx == self.abs_targ_obj_idx
+        picked_up_obj = is_holding_obj and not self._prev_holding_obj
+
+        # DISTANCE REWARD: Steers the agent towards the object and then towards the goal
 
-        is_holding_obj = self._sim.grasp_mgr.snap_idx == abs_targ_obj_idx
         if is_holding_obj:
-            dist = to_goal[targ_obj_idx]
-            dist_diff = (
-                self._prev_measures[1][targ_obj_idx] - to_goal[targ_obj_idx]
-            )
+            dist_diff = self._prev_obj_to_goal_dist - obj_to_goal_dist
         else:
-            dist = to_obj[targ_obj_idx]
-            dist_diff = (
-                self._prev_measures[0][targ_obj_idx] - to_obj[targ_obj_idx]
-            )
+            dist_diff = self._prev_ee_to_obj_dist - ee_to_obj_dist
+        self._metric += self._config.dist_reward * dist_diff
 
-        if (
-            is_holding_obj
-            and not self._prev_holding_obj
-            and self._cur_rearrange_step not in self._did_give_pick_reward
-        ):
+        # PICK REWARD: Reward for picking up the object, only given once to avoid
+        # reward hacking.
+
+        already_gave_reward = (
+            self._cur_rearrange_stage in self._gave_pick_reward
+        )
+        if picked_up_obj and not already_gave_reward:
             self._metric += self._config.pick_reward
-            self._did_give_pick_reward[self._cur_rearrange_step] = True
+            self._gave_pick_reward[self._cur_rearrange_stage] = True
 
-        if (
-            dist < self._config.success_dist
-            and not is_holding_obj
-            and self._cur_rearrange_step < num_targs
-        ):
-            self._metric += self._config.single_rearrange_reward
-            self._cur_rearrange_step += 1
-            self._cur_rearrange_step = min(
-                self._cur_rearrange_step, num_targs - 1
-            )
+        # PLACE REWARD: Reward for placing the object correcly (within success dist)
+        # We also udate the target object for the next stage.
 
-        self._metric += self._config.dist_reward * dist_diff
-        self._prev_measures = (to_obj, to_goal)
+        place_success = obj_to_goal_dist < self._config.success_dist
+        if place_success and not is_holding_obj:
+            self._metric += self._config.single_rearrange_reward
```

Comment:
@SergioArnaud Is it possible to double count the place reward here if the task does not end after placing? I would add a `if self._cur_rearrange_stage < self.num_targets` to make sure of it.

Reply:
@vincentpierre I think that case is considered in [this](https://github.com/vincentpierre/habitat-lab/blob/900629559acec863a70782b7372e21b77e5d3870/habitat-lab/habitat/tasks/rearrange/multi_task/composite_sensors.py#L131-L133) condition. If all the objects are placed in the right place but for some reason we haven't succeded in the task (for example the agent hasn't called terminate) we give zero reward.

Reply:
Oh, nice. sorry I missed it.